session.go

```diff
@@ -259,158 +276,249 @@ func (sess *session) Break(c chan<- bool) {
+				// accepting the session.
+				if sess.sessReqCb != nil && !sess.sessReqCb(sess, req.Type) {
+					sess.rawCmd = ""
+					req.Reply(false, nil)
+					continue
+				}
+
+				handler := sess.subsystemHandlers[payload.Value]
+				if handler == nil {
+					handler = sess.subsystemHandlers["default"]
+				}
+				if handler == nil {
+					req.Reply(false, nil)
+					continue
+				}
+
+				sess.handled = true
+				req.Reply(true, nil)
+
+				go func() {
+					handler(sess)
+					sess.Exit(0)
+				}()
+			case "env":
+				if sess.handled {
+					req.Reply(false, nil)
+					continue
+				}
+				var kv struct{ Key, Value string }
+				gossh.Unmarshal(req.Payload, &kv)
+				sess.env = append(sess.env, fmt.Sprintf("%s=%s", kv.Key, kv.Value))
+				req.Reply(true, nil)
+			case "signal":
+				var payload struct{ Signal string }
+				gossh.Unmarshal(req.Payload, &payload)
+				sess.Lock()
+				if sess.sigCh != nil {
+					sess.sigCh <- Signal(payload.Signal)
+				} else {
+					if len(sess.sigBuf) < maxSigBufSize {
+						sess.sigBuf = append(sess.sigBuf, Signal(payload.Signal))
+					}
+				}
+				sess.Unlock()
+			case "pty-req":
+				if sess.handled || sess.pty != nil {
+					req.Reply(false, nil)
+					continue
+				}
+				ptyReq, ok := parsePtyRequest(req.Payload)
 				if !ok {
 					req.Reply(false, nil)
 					continue
 				}
-			}
-			sess.pty = &ptyReq
-			sess.winch = make(chan Window, 1)
-			sess.winch <- ptyReq.Window
-			defer func() {
-				// when reqs is closed
-				close(sess.winch)
-			}()
-			req.Reply(ok, nil)
-		case "x11-req":
-			if sess.handled || sess.x11 != nil {
-				req.Reply(false, nil)
-				continue
-			}
-			x11Req, ok := parseX11Request(req.Payload)
-			if !ok {
+				if sess.ptyCb != nil {
+					ok := sess.ptyCb(sess.ctx, ptyReq)
+					if !ok {
+						req.Reply(false, nil)
+						continue
+					}
+				}
+				sess.pty = &ptyReq
+				sess.winch = make(chan Window, 1)
+				sess.winch <- ptyReq.Window
+				defer func() {
+					// when reqs is closed
+					close(sess.winch)
+				}()
+				req.Reply(ok, nil)
+			case "x11-req":
+				if sess.handled || sess.x11 != nil {
+					req.Reply(false, nil)
+					continue
+				}
+				x11Req, ok := parseX11Request(req.Payload)
+				if !ok {
+					req.Reply(false, nil)
+					continue
+				}
+				sess.x11 = &x11Req
+				if sess.x11Cb != nil {
+					ok := sess.x11Cb(sess.ctx, x11Req)
+					req.Reply(ok, nil)
+					continue
+				}
 				req.Reply(false, nil)
-				continue
-			}
-			sess.x11 = &x11Req
-			if sess.x11Cb != nil {
-				ok := sess.x11Cb(sess.ctx, x11Req)
+			case "window-change":
+				if sess.pty == nil {
+					req.Reply(false, nil)
+					continue
+				}
+				win, _, ok := parseWindow(req.Payload)
+				if ok {
+					sess.pty.Window = win
+					sess.winch <- win
+				}
 				req.Reply(ok, nil)
-				continue
-			}
-			req.Reply(false, nil)
-		case "window-change":
-			if sess.pty == nil {
+			case agentRequestType:
+				// TODO: option/callback to allow agent forwarding
```

Comment:
nit: create follow-up issue and add link

Reply:
Oh, I must have reformatted the code as it isn't my TODO. I can create an issue though.

Reply:
Ah then please diregard!